EngineLayer/ModernSearch/ModernSearchEngine.cs

```diff
@@ -126,10 +87,19 @@ protected override MetaMorpheusEngineResults RunSpecific()
             return new MetaMorpheusEngineResults(this);
         }
 
-        protected List<int> GetBinsToSearch(Ms2ScanWithSpecificMass scan, List<int>[] FragmentIndex, DissociationType dissociationType)
+        /// <summary>
+        /// This is a first-pass scoring method which is supposed to be *very fast* but may sometimes miscount the number of truly matched fragments. The number
+        /// of matched fragments should always be overestimated, never underestimated.
+        /// </summary>
+        protected void IndexScoreScan(Ms2ScanWithSpecificMass scan, byte[] scoringTable, byte byteScoreCutoff, List<int> peptidesPossiblyObserved, DissociationType dissociationType)
         {
-            int obsPreviousFragmentCeilingMz = 0;
-            List<int> binsToSearch = new List<int>();
```

Comment:
this is just a guess, but I think line 132 of the old code may be why modern search is currently so slow. it's allocating a new list for every fragment in the data file. these lists are collected by the garbage collector, which stops all threads to perform garbage collection.

Reply:
It seems like to me the take(10) helps a lot. Not sure about this, you use "List<int> bin = FragmentIndex[fragmentBin];" in the new method.

Reply:
I was wrong - the current code is allocating a new list for each scan, not for each fragment.

to test my hypothesis I inserted a `List<int> temp = new List<int>();` onto line 96 and added each valid bin to the list.

without new list: 127 sec
with new list, bins added to it: 127 sec

so that doesn't seem like it's the culprit.

I also tested how much time the "top 10" filter saved:
with top 10 filter: 127 sec
without filter: ~2625 sec (estimated based on 12% of the search completing in 315 sec)

So it's definitely the fine-scoring part that's slow. Based on my CPU usage (it's spikey instead of flat) I'm guessing it's because of frequent garbage collection. I'll see if I can optimize this part...

Reply:
I think it is reasonable to only fine score top N. 
In Crosslink search, the alpha peptide always ranks within the top 50. You can output the rank to see if this is the truth.

Reply:
I replaced the top N with a function which always accurately computes the best peptide score (it fine-scores all the peptides that could be the best-scoring), but the delta-score is only an approximation because it takes too long to score all of the peptides that could be second-best scoring. 

because the delta score is only approximated, and PEP analysis trains on delta score, the PEP values are different for classic and modern search. and because some ambiguity is removed based on PEP, some ambiguity is also slightly different. I think it's not an important degree of difference, though.

Reply:
speed on my benchmark test is maintained at 127sec, btw, so this method is just as fast as re-scoring the top 10 and is more accurate.

Reply:
building in release took ~60 seconds... 😵